### PR TITLE
Add client to UFS bridge

### PR DESCRIPTION
**Pull request by LuQQiu**
### What changes are proposed in this pull request?
PR 2 Alluxio FUSE SDK https://github.com/Alluxio/alluxio/pull/16284
Support client connecting to UFS directly without the integration with Alluxio master.

### Why are the changes needed?

Support Alluxio Fuse SDK which connects to UFS directly

### Does this PR introduce any user facing changes?
NA

### CODE REVIEW

**Comment by LuQQiu** [comment]
This PR will focus on the general Client to UFS bridge, and will use another two PRs to do the refactoring jobs
- FileSystemOptions to replace newly added property key and define all FileSystem option passing sequence
- Refactor UnderFileSystem to directly return FileInStream/FileOutStream instead of InputStream/OutputStream

**Comment by LuQQiu** [comment]
@beinan @jja725 PTAL, thanks!

**Review by jja725 — COMMENTED** [review]
So a general comment is that can we get rid of most of the IOException? We should start the new exception convention for the new code. Let me know if you need help on that.
And also couple of early comments since I haven't looked at the implementation detailed yet.

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+  /**
+   * Constructs a new base file system.
+   *
+   * @param fsContext file system context
+   */
+  public UfsBaseFileSystem(FileSystemContext fsContext) {
+    mFsContext = fsContext;
+    String ufsAddress = mFsContext.getClusterConf().getString(PropertyKey.USER_UFS_ADDRESS);
+    mRootUFS = new AlluxioURI(ufsAddress);
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:91]: GOOD CALL, move it under the check

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        new AlluxioURI(ufsAddress));
+    mUfs = ufsClient.acquireUfsResource();
+    mCloser.register(mFsContext);
+    mCloser.register(mUfs);
+    LOG.debug("Creating file system connecting to ufs address {}", ufsAddress);
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:104]: Still valid, verified, thanks

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+   */
+  @Override
+  public synchronized void close() throws IOException {
+    // TODO(zac) Determine the behavior when closing the context during operations.
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:108]: removed

**Comment by LuQQiu** [comment]
@jja725 Move from IOException to Alluxio UFS runtime exception.
For details about how to translate UFS IOException to runtime exception, will be done in other PRs.
The optimal way still be let UFS directly throw AlluxioRuntimeException

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
+   */
+  @Override
+  public synchronized void close() throws IOException {
+    if (!mClosed) {
+      mClosed = true;
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:111]: should we add a return statement here?  I'm just thinking there might be an issue if we call mCloser.close() multiple times.

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+      String dstPath = dst.getPath();
+      boolean renamed;
+      if (mUfs.get().isFile(srcPath)) {
+        renamed = mUfs.get().renameFile(srcPath, dstPath);
+      } else {
+        renamed = mUfs.get().renameDirectory(srcPath, dstPath);
+      }
+      if (!renamed) {
+        throw new AlluxioRuntimeException(
+            Status.UNKNOWN,
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:300]: is it possible to get the error message from ufs?

Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:300]: or at lease we could put the status to "External"?  because it's not our fault.

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Supports reading from a ufs file directly.
+ */
+// TODO(lu) make it thread safe
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java:25]: mostly it's ok if the input stream is not thread-safe

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+@NotThreadSafe
+public class UfsFileInStream extends FileInStream {
+  /** Used to manage closeable resources. */
+  private final Closer mCloser;
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java:29]:
```suggestion
  private final Closer mCloser = Closer.create();
```
and remove line 41

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+   * Creates a new {@link UfsFileInStream}.
+   *
+   * @param stream the embedded input stream
+   * @param fileLength the file length
+   */
+  public UfsFileInStream(InputStream stream, long fileLength) {
+    mCloser = Closer.create();
+    mUfsInStream = Preconditions.checkNotNull(stream);
+    mLength = fileLength;
+    mPosition = 0;
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java:44]: unnecessary code.  the default value of long is 0

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    int res = mUfsInStream.read();
+    if (res > 0) {
+      mPosition += res;
+    }
+    return res;
+  }
+
+  @Override
+  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
+    byte[] byteArray = new byte[len];
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java:59]: there might be performance penalty for the memory allocating and copying.   I would try to make the read to fill the data to byteBuffer directly.   This might be out of scope of this PR, we could do it in a separate PR

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileOutStream.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  private final Closer mCloser;
+  private final OutputStream mUfsOutStream;
+
+  /**
+   * Creates a {@link UfsFileOutStream}.
+   *
+   * @param stream the embedded UFS output stream
+   */
+  public UfsFileOutStream(OutputStream stream) {
+    mCloser = Closer.create();
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/UfsFileOutStream.java:36]: move to line 27

**Review by beinan — COMMENTED** [review]
just a couple of minor issues.   any tests for UfsFileInStream and UfsFileOutStream?

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java (hunk 5 above)
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:300]: Ideally UFS should directly throw AlluxioRuntimeExceptions because there may be different reasons, let me try to synchronize some common error message in this PR
e.g. source does not exist or is directory

Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:300]: One future change is in the TODO, let UFS directly throw detailed runtime exceptions instead of returning a boolean with the only error message in the logs.

Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java:300]:
```
            Status.UNKNOWN,
            String.format("Failed to rename from %s to %s", srcPath, dstPath),
            null,
            ErrorType.External,
            false
        );
```
Could not find the status.External, we are using gRPC status.
But set error type to be external

**Review by jja725 — COMMENTED** [review]
Mostly LGTM, leave some minor comments.

File: core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+      if (options.hasRecursive()) {
+        ufsOptions.setRecursive(options.getRecursive());
+      }
+      mUfs.get().deleteDirectory(ufsPath, ufsOptions);
+    });
+  }
+
+  @Override
+  public boolean exists(AlluxioURI path, final ExistsPOptions options) {
+    return Boolean.TRUE.equals(callWithReturn(() -> mUfs.get().exists(path.getPath())));
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsBaseFileSystem.java]: because the `callWithReturn` returns object instead of primitive types.
So it would return Boolean, already we know Boolean here will not be null, but IDE will show errors that unsafe unboxing

File: core/client/fs/src/main/java/alluxio/client/file/UfsFileOutStream.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+   *
+   * @param stream the embedded UFS output stream
+   */
+  public UfsFileOutStream(OutputStream stream) {
+    mUfsOutStream = Preconditions.checkNotNull(stream);
+    mCloser.register(mUfsOutStream);
+  }
+
+  @Override
+  public synchronized long getBytesWritten() {
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/UfsFileOutStream.java]: e.g. 
Fuse.create() creates the out stream
Fuse.write() write to out stream
...
Fuse.getAttr() wants to know how much bytes are written
Fuse.write()
Fuse.getAttr()
Fuse.close()

So I could keep the information on the FUSE side but if the client side already has such information, it would be easier to maintain and get.
Ideally the underlying UFS can directly tell us bytesWritten, but normally the bytesWritten given by UFSes are incorrect (because of buffering related stuff)

**Comment by LuQQiu** [comment]
@beinan @jja725 Thanks for the review, addressed the comments, PTAL, thanks

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    mUfsInStream.mark(0);
+  }
+
+  @Override
+  public int read() throws IOException {
+    int res = mUfsInStream.read();
+    if (res == -1) {
+      return -1;
+    }
+    mPosition++;
```
Comment (jja725) [inline core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java:55]: might be +=res?

**Review by jja725 — CHANGES_REQUESTED** [review]

**Review by jja725 — APPROVED** [review]

**Comment by beinan** [comment]
alluxio-bot, merge this please.
